pkg/autoupgrade/tags.go

```diff
@@ -61,12 +61,12 @@ func findLatestTagForImageWithPattern(ctx context.Context, c daemonClient, names
 		}
 	}
 
-	return newImage, newTag != imagename.DefaultTag, err
+	return newImage, newTag != current, err
```

Comment:
Im starting to think that there is a more fundamental flaw here. This function is used in two places: the ImageDetails validator and the autoUpgrade daemon. The two callers have different expectations of what the boolean value returned is supposed to mean. The imageDetails caller thinks it means "a matching tag was found". The daemon thinks it means "I found an image you should update to because it is better than your current". They're similar but subtly different.

So, now I don't think passing in the "current" tag is the right thing to do (because that doesnt jive with the usage in imageDetails).

I'm still thinking this through

cc @thedadams

Reply:
Ok, put together a thought on how to tackle this.

First, the boolean return by this function should be defined to mean whether a positive matching tag has been found. We should not pass "current" into this function and attempt to compare here.

Secondly (and somewhat unrelated) `newTag != imagename.DefaultTag` seems like it could introduce a subtle bug to me because `imagename.DefaultTag` ("latest") is a valid tag that could sometimes be the right choice as the latest tag. So, higher up in this function, instead of using imagename.DefaultTag as the placeholder for "a tag cant be found", use a different value that would actually be an invalid tag like `"+notfound+"` (`+` is an illegal character).

Thirdly, in the daemon's use case, you do need to compare the image returned by this function to the app's current image, but just do it there, after you return from this function. Don't try to do it in this function.

@thedadams feel free to chime in on this.

Reply:
Yes, I agree with everything you said here. I will add two thoughts.

Based on what you said, the best way to **guarantee** that we haven't found a matching image is to use something like `+notfound+` as the "starting tag" because we know that will never be a valid matching tag.

Secondly, the use of `+notfound+` is an implementation detail of this function and shouldn't leak out. Meaning that, even if `newTag == "+notfound+"`, the `newImage` returned here shouldn't contain that as its tag. Since `+notfound+` is not a valid tag, returning an image with that tag may produce bugs/errors in other areas, too.

Reply:
I agree with the first 2 points of Craig as well and i will go ahead with that implementation change as recommended by @thedadams.  👍 

Out of curiosity I was wondering if it would have been better to have a function present in the `tags.go` which would either use `findLatestTagForImageWithPattern` or use similar functionality with added use of current to compare with the latest image. Essentially to handle what has been said before - The imageDetails caller thinks it means "a matching tag was found". The daemon thinks it means "I found an image you should update to because it is better than your current".

Reply:
I had a similar thought. I even wrote up an entire expose on doing something like that.

The reason I didn't post it is the name `findLatestTagForImageWithPattern` implies (to me) that it will return the latest tag. It would be up to the caller to determine what to do with the latest tag.

Reply:
This is looking good. I'm going to resolve this convesation.